acpi/src/lib.rs

```diff
@@ -115,170 +138,150 @@ where
-            }
-            Signature::SSDT => {
-                self.ssdts.push(AmlTable::new(physical_address, header.length));
+    /// Finds and returns the DSDT AML table, if it exists.
+    pub fn dsdt(&self) -> AcpiResult<AmlTable> {
+        self.find_table::<fadt::Fadt>().and_then(|fadt| {
+            struct Dsdt;
+            // ### Safety: Implementation properly represents a valid DSDT.
+            unsafe impl AcpiTable for Dsdt {
+                const SIGNATURE: Signature = Signature::DSDT;
+
+                fn header(&self) -> &sdt::SdtHeader {
+                    // ### Safety: DSDT will always be valid for an SdtHeader at its `self` pointer.
+                    unsafe { &*(self as *const Self as *const sdt::SdtHeader) }
+                }
             }
-            signature => {
-                self.sdts.insert(signature, Sdt { physical_address, length: header.length, validated: false });
-            }
-        }
 
-        Ok(())
+            let dsdt_address = fadt.dsdt_address()?;
+            let dsdt = unsafe { read_table::<H, Dsdt>(self.handler.clone(), dsdt_address)? };
+
+            Ok(AmlTable::new(dsdt_address, dsdt.header().length))
+        })
     }
 
-    /// Create a mapping to a SDT, given its signature. This validates the SDT if it has not already been
-    /// validated.
-    ///
-    /// ### Safety
-    /// The table's memory is naively interpreted as a `T`, and so you must be careful in providing a type that
-    /// correctly represents the table's structure. Regardless of the provided type's size, the region mapped will
-    /// be the size specified in the SDT's header. Providing a `T` that is larger than this, *may* lead to
-    /// page-faults, aliasing references, or derefencing uninitialized memory (the latter two of which are UB).
-    /// This isn't forbidden, however, because some tables rely on `T` being larger than a provided SDT in some
-    /// versions of ACPI (the [`ExtendedField`](crate::sdt::ExtendedField) type will be useful if you need to do
-    /// this. See our [`Fadt`](crate::fadt::Fadt) type for an example of this).
-    pub unsafe fn get_sdt<T>(&self, signature: sdt::Signature) -> Result<Option<PhysicalMapping<H, T>>, AcpiError>
-    where
-        T: AcpiTable,
-    {
-        let sdt = match self.sdts.get(&signature) {
-            Some(sdt) => sdt,
-            None => return Ok(None),
-        };
-        let mapping = unsafe { self.handler.map_physical_region::<T>(sdt.physical_address, sdt.length as usize) };
-
-        if !sdt.validated {
-            mapping.header().validate(signature)?;
+    /// Iterates through all of the SSDT tables.
+    pub fn ssdts(&self) -> SsdtIterator<H> {
+        let table_base_address = unsafe { self.mapping.virtual_start().as_ptr().add(1) as usize };
+        let table_end_address = table_base_address
+            + (((self.mapping.length as usize) - mem::size_of::<SdtHeader>()) / mem::size_of::<u64>());
```

Comment:
(potentially dumb question - it's 1am here) why do we divide by 8 to get the end-address? Wouldn't this be calculating the number of tables and then adding it to the start address?

Reply:
I believe the table is laid out as a sequence of 64bit-wide pointers? I'm admittedly not certain on this one either; I'll consult the spec.

Reply:
It is, but my point is don't you need the total number of bytes if you want the end address, rather than dividing to get the number of tables? As in shouldn't this be just:
``` rust
let table_end_address = table_base_address + (self.mapping.length as usize) - mem::size_of::<SdtHeader>();
```

Reply:
You're right. I've gone ahead and rewritten the `SsdtIterator` to function a little more idiomatically.